### ISSUE TITLE
guidance on use of JSON-LD?

### ISSUE THREAD

**Issue by dbaron** [w3ctag/design-principles#128]
I think we should consider giving guidance on the use of JSON-LD.  In particular, it seems to me that JSON-LD documents can be interpreted in two ways, either (1) as a JSON document or (2) based on the data model defined in the JSON-LD spec.  It's not clear to me if there are clearly-defined terms for these two things, though.

Specifications that use JSON-LD ought to be clear when users of the JSON-LD involved must interpret it in one of the two ways, or when they're free to use either.  (I could imagine the answer might vary between specifications, or might vary between conformance classes in a specification.)

I'd note that I prefer that browsers always do (1).

**Comment by hadleybeeman** [comment]
@dbaron, we've discussed this on today's TAG breakout. While it doesn't look like many (if any?) browsers currently parse JSON-LD with its data model - I worry about explicitly say that they won't or shouldn't. @torgo wondered if we could pull together some language about how a spec authory may not be able to expect the browser to parse it... But we then brainstormed a number of different things that a spec author may not expect the browser to do, and were struggling to work out why parsing JSON-LD with its data model might be worth calling out. What are your thoughts here? What would be the thing you'd hope we could solve by adding something about this to the Design Principles doc?

**Comment by marcoscaceres** [comment]
TAG should work with the editors of JSON-LD to define the actual data model of JSON-LD in terms of WHATWG Infra. I think the current spec is potentially harmful because it tries to naively piggyback on WebIDL without consideration that WebIDL is very browser specific.  

This is potentially harmful to JSON-LD (and any organization adopting it) in the long term, because JSON-LD wants to be language/implementation agnostic, which WebIDL is not: WebIDL is very much bound to how browsers work. There is no WebIDL implementation in Ruby, or even completely in JavaScript, for instance - that should be of concern. 

An alternative might be to suggest that JSON-LD be expressed in terms of WHATWG Infra types + use Infra's algorithms for converting JSON into infra types. That would give a good language/implementation agnostic data model, plus allow JSON-LD to be treated truly as JSON. The browsers don't need to care because they support JSON (addressing @dbaron's concern (1)) and it addresses (2), in that Infra types map 1-to-1 to JavaScript types - as well as C++ types, and Rust, and basically any other language that supports primitive abstract types (strings, lists, numbers, booleans, bytes, etc.).

**Comment by torgo** [comment]
Discussed in breakout [at NZ F2f](https://github.com/w3ctag/meetings/blob/gh-pages/2020/03-wellington/05-03-minutes.md).

**Comment by annevk** [comment]
https://hsivonen.fi/no-json-ns/ seems relevant here. It's really not clear to me we should be endorsing JSON-LD.

**Comment by marcoscaceres** [comment]
I'm also disappointed that my concerns are dismissed as a "clash of communities" - they are not. The spec being defined in terms of WebIDL makes it problematic for the technical reasons I outlined. I've discussed that with the JSON-LD folks also. Using WebIDL to define the types gives them a particular meaning and behavior (as per WebIDL) - using WebIDL naively risk causing significant issues. 

To put it differently, I don't mind that the W3C community defines data models and formats, but we should do it properly and use things for their intended purposes.

**Comment by othermaciej** [comment]
JSON LD API 1.1 has IDL snippets that claim interfaces are to be exposed on Window and Worker. This does not make any sense for an API that's not intended to be exposed in the browser. 

For example [RdfDataset](https://www.w3.org/TR/json-ld11-api/#rdf-dataset-interfaces):

It's not clear how this could have gotten to PR without browser-hosted implementations, since `Window` and `Worker` only exist in browsers.

```
[Exposed=(Window,Worker)]
interface RdfDataset {
  constructor();
  readonly attribute RdfGraph defaultGraph;
  void add(USVString graphName, RdfGraph graph);
  iterable<USVString?, RdfGraph>;
};
```

**Comment by gkellogg** [comment]
> JSON LD API 1.1 has IDL snippets that claim interfaces are to be exposed on Window and Worker. This does not make any sense for an API that's not intended to be exposed in the browser.

Note that the [editors draft](https://w3c.github.io/json-led-api) has been updated to remove the Exposed declaration. But really, IMO, what is needed is an alternate way to expose APIs for non-browser APIs, which are arguably a bigger use case than browser-specific use cases.
> It's not clear how this could have gotten to PR without browser-hosted implementations, since `Window` and `Worker` only exist in browsers.

Advice earlier was to use that, as WebIDL requires something be exposed. There are comments in the [API preamble](https://w3c.github.io/json-ld-api/#the-application-programming-interface) that describes the purpose of the API and that implementations should implement as appropriate for their environments.

The fact is, JSON-LD started using WebIDL in 2014 with the [1.0 spec](https://www.w3.org/TR/2014/REC-json-ld-api-20140116/#the-application-programming-interface) and things seem to have changed with the community since then, to the exclusion of these use cases. This is wrongheaded, in my mind, and the W3C needs a single way to normatively express APIs that aren’t narrowly focused on the needs of browsers.

**Comment by hober** [comment]
> Note that the [editors draft](https://w3c.github.io/json-led-api) has been updated to remove the Exposed declaration. But really, IMO, what is needed is an alternate way to expose APIs for non-browser APIs, which are arguably a bigger use case than browser-specific use cases.

Here's an idea that just popped into my head that I haven't thought through. Why not define your own global for all of your interfaces?

```
[Global=Something,
 Exposed=Something]
interface Something : EventTarget {
    ...
}
```

Then you can say `[Exposed=Something]` on your interfaces. This would be valid WebIDL and it would make no claim to browser implementation.

**Comment by gkellogg** [comment]
As a workaround, that seems reasonable, although I believe something should be defined.

However, trying it in ReSpec gives me the following errors:

```
[Global=Endpoint, Exposed=Endpoint]
interface Endpoint : JsonLdProcessor {
...
}

WebIDL validation error: Interfaces marked as [Global] cannot have constructors ...
WebIDL validation error: The name "Endpoint" of type "interface" was already seen ...
```

But, of course, there are probably aspects of WebIDL that are escaping me. I'd certainly appreciate some more specific help on the repo, but it may need to be something that is taken up by the proposed maintenance WG post-publication.

**Comment by hober** [comment]
> ```
> [Global=Endpoint, Exposed=Endpoint]
> interface Endpoint : JsonLdProcessor {
> ...
> }
> 
> WebIDL validation error: Interfaces marked as [Global] cannot have constructors ...
> WebIDL validation error: The name "Endpoint" of type "interface" was already seen ...
> ```

Err, yeah, that's not gonna do what you want. You're defining Endpoint as a subclass of JsonLdProcessor, which is not what you want.

What about something like this?

```
[Global=Thingy, Exposed=Thingy]
interface Thingy {
}
```

And then in all of your other interface definitions, you say `[Exposed=Thingy]` like so:

```
[Exposed=Thingy]
interface RdfDataset {
  constructor();
  ...
};

**Comment by gkellogg** [comment]
That seems to do it, if it is generally acceptable. I also need to remove the `constructor()`, as a global can't have a constructor.

I created a [PR for this](https://github.com/w3c/json-ld-api/pull/500), and fortunately, PR Preview is working, so you can view a [Preview](https://pr-preview.s3.amazonaws.com/w3c/json-ld-api/pull/500.html#the-application-programming-interface) and [Diffs](https://pr-preview.s3.amazonaws.com/w3c/json-ld-api/500/e4e8dc5...879643f.html#the-application-programming-interface).

**Comment by marcoscaceres** [comment]
I still think this is working around the fundamental problem that WebIDL shouldn't be used for this at all.

**Comment by gkellogg** [comment]
> I still think this is working around the fundamental problem that WebIDL shouldn't be used for this at all.

There are many non-browser specs which have a need to describe interfaces, and the narrow focus of WebIDL on browser vendors does not serve that audience. The fact is, the tooling has built-in support for WebIDL, and it is the de-facto standard for describing interfaces at the W3C. If we didn't have that, we'd need to invent something else, which likely heavily leverage WebIDL anyway.

The fact that many implementations have been able to use the JSON-LD API, as specified by WebIDL and create broadly interoperable implementations shows that WebIDL is effective at describing these interfaces, even if specifics such as Promises become more of a useful abstraction than an actuality.

IMO, the charter of WebIDL needs to be opened to server the larger needs for spec editors at the W3C. @hober's advice on using globals should make it clear that the interfaces are not expected to be implemented in browsers, but I think it's a stop-gap solution before the technology faces this need head-on.

**Comment by hober** [comment]
> There are many non-browser specs which have a need to describe interfaces, and the narrow focus of WebIDL on browser vendors does not serve that audience. […] @hober's advice on using globals should make it clear that the interfaces are not expected to be implemented in browsers, but I think it's a stop-gap solution before the technology faces this need head-on.

Yes, my suggestion is a stop-gap. In the longer run, I think you'd be much better off defining things in terms of the WHATWG Infra Standard, like @marcoscaceres suggested in February:

> TAG should work with the editors of JSON-LD to define the actual data model of JSON-LD in terms of WHATWG Infra. I think the current spec is potentially harmful because it tries to naively piggyback on WebIDL without consideration that WebIDL is very browser specific.
> 
> This is potentially harmful to JSON-LD (and any organization adopting it) in the long term, because JSON-LD wants to be language/implementation agnostic, which WebIDL is not: WebIDL is very much bound to how browsers work. There is no WebIDL implementation in Ruby, or even completely in JavaScript, for instance - that should be of concern.
> 
> An alternative might be to suggest that JSON-LD be expressed in terms of WHATWG Infra types + use Infra's algorithms for converting JSON into infra types. That would give a good language/implementation agnostic data model, plus allow JSON-LD to be treated truly as JSON. The browsers don't need to care because they support JSON (addressing @dbaron's concern (1)) and it addresses (2), in that Infra types map 1-to-1 to JavaScript types - as well as C++ types, and Rust, and basically any other language that supports primitive abstract types (strings, lists, numbers, booleans, bytes, etc.).

**Comment by gkellogg** [comment]
> Yes, my suggestion is a stop-gap. In the longer run, I think you'd be much better off defining things in terms of the WHATWG Infra Standard, like @marcoscaceres suggested in February:

We do define our datatypes in terms of Infra (e.g., [default object](https://www.w3.org/TR/json-ld11/#dfn-default-object) is defined in terms of [map](https://infra.spec.whatwg.org/#ordered-map). What this does not provide us is a way of actually defining the calling interfaces, exceptions, options, and so forth. We define a number of different entry points that have specific defined behavior, and this does not seem to be use case handled by Infra. We also take advantage of the ReSpec support for WebIDL to reference methods, arguments, options, and errors which is quite useful and it seems arbitrary to not avail ourselves, and implementors, of these cross-linking and formatting capabilities, or would require each editor to re-create a mechanism for doing this in an arbitrary manner.

**Comment by hadleybeeman** [comment]
@hober and I looked at this in our virtual W3CTAG face-to-face.

The discussion about webIDL deserves its own issue, so we'd like to move it over to [issue 216](https://github.com/w3ctag/design-principles/issues/216).

**Comment by dbaron** [comment]
@hober and @hadleybeeman asked me (during the TAG's plenary meeting during our virtual face-to-face after sessions 15 and 16) to have a look at this after their discussion today, and wanted me to confirm that now that both #216 and #239 exist, it's ok to close this issue.  I agree that those adequately cover the issue I was raising and I think also the issue that @marcoscaceres raised, so I agree that this issue can now be closed in favor of those issues.
